ass5.cpp

```diff
@@ -0,0 +1,117 @@
+#include <iostream>
+#include <vector>
+#include <stack>
+
+using namespace std;
+
+#define ALPHABET_SIZE 50
+
+class Graph {
+    vector<vector<int> > adj_list;
+    
+public:
+    Graph () {
+        adj_list.resize(ALPHABET_SIZE);
+    }
+    
+    void addEdge(int v, int w) {
+        adj_list[v].push_back(w);
+    }
+    
+    void topologicalSortUtil(int v, vector<bool> &visited, stack<int> &stack);
+    
+    stack<int> topologicalSort(vector<int> letter_ids);
+};
+
+void Graph::topologicalSortUtil(int v, vector<bool> &visited, stack<int> &stack) {
+    visited[v] = true;
+    
+    for (auto i : adj_list[v]) {
+        if (!visited[i]) {
+            topologicalSortUtil(i, visited, stack);
+        }
+    }
+    
+    stack.push(v);
+}
+
+stack<int> Graph::topologicalSort(vector<int> letter_ids) {
+    stack<int> stack;
+    vector<bool> visited(letter_ids.size(), false);
+
+    for (auto i: letter_ids) {
+        if (!visited[i]) {
+            topologicalSortUtil(i, visited, stack);
+        }
+    }
+    
+    return stack;
+}
+
+string return_alphabet(vector<string> words) {
+    Graph graph;
+    vector<int> letter_ids;
+    
+    for (int i = 0; i < words.size() - 1; ++i) {
+        string str_first = words[i];
+        string str_sec = words[i + 1];
+        
+        for (int j = 0; j < min(str_first.length(), str_sec.length()); ++j) {
+            if (str_first[j] != str_sec[j]) {
+                graph.addEdge(str_first[j] - 'A', str_sec[j] - 'A');
```

Comment:
Also, char can be signed or unsigned, depending on the compiler/platform. This may generate a hard to debug bug, that works on your machine, but not on production.